File: backend/internal/handler/projects.go
package handler

import (
	"github.com/gofiber/fiber/v2"
)

type ProjectBody struct {
	Name        string `json:"name"`
	Path        string `json:"path"`
	TestCommand string `json:"testCommand"`
	PipelineID  string `json:"pipelineId"`
	Description string `json:"description"`
}

func (h *Handler) ListProjects(c *fiber.Ctx) error {
	rows, err := h.db.Query(c.Context(), `
		SELECT pr.id, pr.name, pr.path, pr.test_command, pr.pipeline_id, pr.description, pr.created_at, pr.updated_at,
		  p.name AS pipeline_name
		FROM projects pr
		JOIN pipelines p ON p.id = pr.pipeline_id
		ORDER BY pr.created_at DESC`)
	if err != nil {
		return fiber.NewError(500, err.Error())
	}
	defer rows.Close()

	var projects []fiber.Map
	for rows.Next() {
		var id, name, path, testCmd, pipelineID, desc, pipelineName string
		var createdAt, updatedAt interface{}
		if err := rows.Scan(&id, &name, &path, &testCmd, &pipelineID, &desc, &createdAt, &updatedAt, &pipelineName); err != nil {
			return fiber.NewError(500, err.Error())
		}
		projects = append(projects, fiber.Map{
			"id": id, "name": name, "path": path, "testCommand": testCmd,
			"pipelineId": pipelineID, "pipelineName": pipelineName,
			"description": desc, "createdAt": createdAt, "updatedAt": updatedAt,
		})
	}
	if projects == nil {
		projects = []fiber.Map{}
	}
	return c.JSON(projects)
}

func (h *Handler) GetProject(c *fiber.Ctx) error {
	id := c.Params("id")
	row := h.db.QueryRow(c.Context(), `
		SELECT pr.id, pr.name, pr.path, pr.test_command, pr.pipeline_id, pr.description, pr.created_at, pr.updated_at,
		  p.name AS pipeline_name
		FROM projects pr JOIN pipelines p ON p.id = pr.pipeline_id
		WHERE pr.id = $1::uuid`, id)
	var rid, name, path, testCmd, pipelineID, desc, pipelineName string
	var createdAt, updatedAt interface{}
	if err := row.Scan(&rid, &name, &path, &testCmd, &pipelineID, &desc, &createdAt, &updatedAt, &pipelineName); err != nil {
		return fiber.NewError(404, "project not found")
	}
	return c.JSON(fiber.Map{
		"id": rid, "name": name, "path": path, "testCommand": testCmd,
		"pipelineId": pipelineID, "pipelineName": pipelineName,
		"description": desc, "createdAt": createdAt, "updatedAt": updatedAt,
	})
}

func (h *Handler) CreateProject(c *fiber.Ctx) error {
	var body ProjectBody
	if err := c.BodyParser(&body); err != nil {
		return fiber.NewError(400, err.Error())
	}
	if body.Name == "" || body.Path == "" || body.PipelineID == "" {
		return fiber.NewError(422, "name, path, pipelineId required")
	}
	row := h.db.QueryRow(c.Context(), `
		INSERT INTO projects (name, path, test_command, pipeline_id, description)
		VALUES ($1,$2,$3,$4::uuid,$5)
		RETURNING id,name,path,test_command,pipeline_id,description,created_at,updated_at`,
		body.Name, body.Path, body.TestCommand, body.PipelineID, body.Description,
	)
	var id, name, path, testCmd, pipelineID, desc string
	var createdAt, updatedAt interface{}
	if err := row.Scan(&id, &name, &path, &testCmd, &pipelineID, &desc, &createdAt, &updatedAt); err != nil {
		return fiber.NewError(500, err.Error())
	}
	return c.Status(201).JSON(fiber.Map{
		"id": id, "name": name, "path": path, "testCommand": testCmd,
		"pipelineId": pipelineID, "description": desc,
		"createdAt": createdAt, "updatedAt": updatedAt,
	})
}

func (h *Handler) UpdateProject(c *fiber.Ctx) error {
	id := c.Params("id")
	var body ProjectBody
	if err := c.BodyParser(&body); err != nil {
		return fiber.NewError(400, err.Error())
	}
	row := h.db.QueryRow(c.Context(), `
		UPDATE projects SET name=$2,path=$3,test_command=$4,pipeline_id=$5::uuid,description=$6,updated_at=NOW()
		WHERE id=$1::uuid
		RETURNING id,name,path,test_command,pipeline_id,description,created_at,updated_at`,
		id, body.Name, body.Path, body.TestCommand, body.PipelineID, body.Description,
	)
	var rid, name, path, testCmd, pipelineID, desc string
	var createdAt, updatedAt interface{}
	if err := row.Scan(&rid, &name, &path, &testCmd, &pipelineID, &desc, &createdAt, &updatedAt); err != nil {
		return fiber.NewError(500, err.Error())
	}
	return c.JSON(fiber.Map{
		"id": rid, "name": name, "path": path, "testCommand": testCmd,
		"pipelineId": pipelineID, "description": desc,
		"createdAt": createdAt, "updatedAt": updatedAt,
	})
}

func (h *Handler) DeleteProject(c *fiber.Ctx) error {
	id := c.Params("id")
	h.db.Exec(c.Context(), "DELETE FROM projects WHERE id=$1::uuid", id)
	return c.SendStatus(204)
}

File: backend-rust/src/repo/projects.rs
use sqlx::PgPool;
use uuid::Uuid;
use crate::domain::project::Project;
use crate::error::{ApiError, ApiResult};

pub struct ProjectRepo;

impl ProjectRepo {
    pub async fn list(pool: &PgPool) -> ApiResult<Vec<Project>> {
        sqlx::query_as::<_, Project>("SELECT * FROM projects ORDER BY created_at DESC")
            .fetch_all(pool)
            .await
            .map_err(|e| e.into())
    }

    pub async fn get(pool: &PgPool, id: Uuid) -> ApiResult<Option<Project>> {
        sqlx::query_as::<_, Project>("SELECT * FROM projects WHERE id = $1")
            .bind(id)
            .fetch_optional(pool)
            .await
            .map_err(|e| e.into())
    }

    pub async fn create(
        pool: &PgPool,
        name: String,
        path: String,
        test_command: Option<String>,
        pipeline_id: Uuid,
        description: Option<String>,
    ) -> ApiResult<Project> {
        sqlx::query_as::<_, Project>(
            "INSERT INTO projects (name, path, test_command, pipeline_id, description, created_at, updated_at)
             VALUES ($1, $2, $3, $4, $5, NOW(), NOW())
             RETURNING *"
        )
        .bind(name)
        .bind(path)
        .bind(test_command.unwrap_or_default())
        .bind(pipeline_id)
        .bind(description.unwrap_or_default())
        .fetch_one(pool)
        .await
        .map_err(|e| e.into())
    }

    pub async fn update(
        pool: &PgPool,
        id: Uuid,
        name: Option<String>,
        path: Option<String>,
        test_command: Option<String>,
        pipeline_id: Option<Uuid>,
        description: Option<String>,
    ) -> ApiResult<Option<Project>> {
        let existing = Self::get(pool, id).await?;
        if existing.is_none() {
            return Ok(None);
        }

        let project = sqlx::query_as::<_, Project>(
            "UPDATE projects SET
             name = COALESCE($1, name),
             path = COALESCE($2, path),
             test_command = COALESCE($3, test_command),
             pipeline_id = COALESCE($4, pipeline_id),
             description = COALESCE($5, description),
             updated_at = NOW()
             WHERE id = $6
             RETURNING *"
        )
        .bind(name)
        .bind(path)
        .bind(test_command)
        .bind(pipeline_id)
        .bind(description)
        .bind(id)
        .fetch_one(pool)
        .await
        .map_err(|e| e.into())?;

        Ok(Some(project))
    }

    pub async fn delete(pool: &PgPool, id: Uuid) -> ApiResult<bool> {
        let result = sqlx::query("DELETE FROM projects WHERE id = $1")
            .bind(id)
            .execute(pool)
            .await
            .map_err(|e| e.into())?;
        Ok(result.rows_affected() > 0)
    }

    pub async fn check_pipeline_usage(pool: &PgPool, pipeline_id: Uuid) -> ApiResult<i64> {
        let result = sqlx::query_scalar::<_, i64>(
            "SELECT COUNT(*) FROM projects WHERE pipeline_id = $1"
        )
        .bind(pipeline_id)
        .fetch_one(pool)
        .await
        .map_err(|e| e.into())?;
        Ok(result)
    }
}
